### PR TITLE
[JENKINS-36283] custom pr behavior

### PR DESCRIPTION
**Pull request by pascallap**
Put in place the bases to solve: [Jenkins-36283](https://issues.jenkins-ci.org/browse/JENKINS-36283)
Option to choose:
- [x] Build origin branches (HEAD) 
- [x] Build origin branches also filed as PRs  (HEAD)
- [x] Build origin/fork PRs (unmerged head) (HEAD)
- [x] Build fork PRs (unmerged head)
- [x] Build origin PRs (merged with base branch)
- [x] Build fork PRs (merged with base branch)

### CODE REVIEW

**Comment by tuhoojabotti** [comment]
This is great, we're currently building all commits twice because all our prs come from local branches.

**Comment by tknerr** [comment]
+1 for this. I guess this is meant to supersede #9 then @amuniz ?

**Comment by Eernie** [comment]
Yes I'd rather see this merged then #9 or #25 imho. Just because it has multiple options and is more like [Github-source-plugin-#60](https://github.com/jenkinsci/github-branch-source-plugin/pull/60).

**Comment by amuniz** [comment]
Thanks for the PR, let's wait until #23 is merged to start reviewing this, as it is changing the `SCMHead` representations and will have impact in your implementation.

I'll probably ping you for minor fixes. If some hard conflict happens after #23 I can take care of it.

**Comment by pascallap** [comment]
Pull Request is know rebase from #23. So conflict should be already resolved.

**Comment by abubeck** [comment]
@amuniz: A merge of this PR would also be very important for us. Let us know if we can contribute for getting this merged.

**Comment by jpigree** [comment]
@amuniz. Same here. This feature will be awesome for us.

Feel free to ask us for help aswell.

**Comment by amuniz** [comment]
I'll try to review this as soon as possible, it is next in my queue.

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMSource.java [hunk 1]
```diff
@@ -372,63 +449,151 @@ private void retrieveBranches(SCMSourceCriteria criteria, @NonNull final SCMHead
...
+                }else if(buildOriginPRHead){
+                    head = new PullRequestSCMHead(owner, repositoryName, branchName, pr, false);
+                    revision = getRevision(head, hash, pr);
+                    observer.observe(head, revision);
+                }else if(buildOriginPRMerge){
+                    head = new PullRequestSCMHead(owner, repositoryName, branchName, pr, true);
+                    revision = getRevision(head, hash, pr);
+                    observer.observe(head, revision);
+                }
+            }
```
Comment (estyrke) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMSource.java]: Suggestion if you'd like to reduce the amount of duplication here:
```
bool buildPRMerge = fork ? buildForkPRMerge : buildOriginPRMerge;
bool buildPRHead = fork ? buildForkPRHead : buildOriginPRHead;
bool addHeadSuffix = buildPRMerge && buildPRHead;

if (buildPRMerge) {
    head = new PullRequestSCMHead(owner, repositoryName, branchName, pr, true);
    revision = getRevision(head, hash, pr);
    observer.observe(head, revision);
}

if (buildPRHead) {
    head = new PullRequestSCMHead(owner, repositoryName, branchName, pr, false, addHeadSuffix);
    revision = getRevision(head, hash, pr);
    observer.observe(head, revision);
}
```

Comment (pascallap) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMSource.java]: Thanks, I knew that there was a better way to handle this.

**Comment by jeffweiss** [comment]
Review ping. This will be a non-trivial improvement in my life. Would love to see it released.

**Comment by raulgd** [comment]
Being able to only build PRs and not branches and PRs at the same time is a big must for us as well, I'd love to get this installed on our Jenkins as soon as possible :)

**Comment by tknerr** [comment]
If there's anything we can help in testing, please let us know.

I was about to build this PR but then noticed that some other SCM related plugins are missing / not published yet in the 2.x release line (see [blog post here](https://jenkins.io/blog/2017/01/17/scm-api-2/))

Does it make sense to take this for a spin now, or better wait for a re-release of the 2.x release line of the SCM plugins?

**Comment by pascallap** [comment]
This Pull Request is using the added objects from 2.x scm api.

If you want to test/evaluate functionality, you will have to use the experimental plugins repository. So you will get access to the 2.x plugin, until they are back in the stable plugin repository. (Don't do it on production system)

For the re-release of the 2.x plugins and the value to review this at the moment, @amuniz and @stephenc are probably the best persons able to answer those questions.

**Comment by luispabon** [comment]
It would be great to allow "only PRs" as well as add specific branches. This would allow us to only build PRs as well as the master branch when things are merged into it.

**Comment by abishek3876** [comment]
@amuniz , any ETA possible for this? Looks like this is a must have for many (including me).

**Comment by stephenc** [comment]
Personally I was to fix the "delete everything if there is an api error" bug first and get that released before considering and additional changes

**Comment by stephenc** [comment]
This really feels like replication of the death by 1000 check-boxes anti-pattern in the GitHub Branch Source. There will likely be other tweaks that will be needed on the PR building behaviour, so I would rather get the behaviour plugable and fix both plugins UI at the same time

**Comment by estyrke** [comment]
@stephenc That would be awesome, but what time frame are we talking then? Me and others have been longing for this functionality for half a year now, and at least for my own part I'd love to have death by checkboxes instead of nothing at all while we wait for a superior alternative.

**Comment by jeffweiss** [comment]
@stephenc this implementation might be death by a 1000 cuts for the handful of plugin developers working in both GU and BB, but the larger number of users of this plugin are already suffering from death by 1000 cuts of duplicate builds for origin PRs.

**Comment by stephenc** [comment]
@jeffweiss I am just stating my opinion and the reason why I am not merging this PR right now.

Top of my action list is fixing GitHub rate limits. Second on my list is fixing event support in this plugin. Third is solving the death by 1000 check boxes in both plugins.

Last week top of my list was api errors deleting everything in a huff, which I claim is fixed in the 2.1.0 release. I am not ignoring this problem, rather trying to clearly state where my priorities are and track addressing the issues I have got approval to work on.

I wouldn't want you to think I am ignoring the effort involved in this PR, rather I want to ensure that the effort gets acknowledged

Even if we end up taking the plugin in a slightly different direction, I will probably borrow some of the changes here... just not the UI changes

**Comment by pascallap** [comment]
@stephenc Thanks for the information.

We thought, base on the comments on other similar PR, that using the same interface/options already present in the Github version was the way to go.

We do agree that it is probably not the best way to present the option. Still it is following what was asked by the main commiter on the project.

If there is anything that we can do to help you, ask and we will see what we can do.

Meanwhile, we will less keep this PR on track and _conflictless_ of the master branch.

**Comment by amuniz** [comment]
> Still it is following what was asked by the main committer on the project.

Right, thanks for that. 

Actually the idea is still the same: get this plugin (in features and UX) as similar as possible to GH Branch Source. The thing is that @stephenc is going to put common configurations in the API plugins (smc-api and branch-api) and fix the checkbox hell in a single place. So :+1: for me.

**Comment by Nossnevs** [comment]
This is somewhat related
https://issues.jenkins-ci.org/browse/JENKINS-42458

**Comment by jeffweiss** [comment]
Review/merge ping. What's the blocker?

**Comment by tknerr** [comment]
+1 been quiet for a while but still eagerly waiting for that! 👍

**Comment by stephenc** [comment]
I am against complicating the code further until I tidy up the configuration options.

**Comment by jeffweiss** [comment]
@stephenc that's incongruent with the priorities you indicated 3 weeks ago.

**Comment by kevcodez** [comment]
Any ETA for merge @stephenc ?

**Comment by fonsecas72** [comment]
Awesome job guys!
I've solved the conflicts here https://github.com/fonsecas72/bitbucket-branch-source-plugin/tree/pullrequest_with_shits as a workaround and everything seems fine (some config labels are not working but should be some mistake I made).
I hope that you can merge this soon.

**Comment by kevcodez** [comment]
@fonsecas72 Is there a workaround to skip building PRs for now? 

This pull request is about 5 months old and still not merged unfortunately.

**Comment by luispabon** [comment]
@kevcodez we have some Jenkinsfile trickery to build either master or PR branches only:

```groovy
def isPr = {
    String branch = env.BRANCH_NAME;
    return branch.startsWith('PR-')
}
```

Then simply

```groovy
node {
    if (env.BRANCH_NAME == 'master' || isPr() == true) {
        // your stuff
    }
```

**Comment by estyrke** [comment]
@kevcodez @luispabon We have the following in "Branch names to build automatically" in the organization folders, which accomplishes the same thing: `(master)|(PR-\d+)`. It should be possible to construct a regex that will select everything except PRs as well (maybe `^(?!PR-\d+$)`?)

**Comment by kevcodez** [comment]
@estyrke 

Looking up TSU/eos-ts-core for branches
Checking branch feature/TZ-1262 from TSU/eos-ts-core
      ‘Jenkinsfile’ found
Met criteria
Looking up TSU/eos-ts-core for pull requests
Checking PR from TSU/eos-ts-core and branch bugfix/TZ-1193
      ‘Jenkinsfile’ found
Met criteria

It does exclude PRs.

**Comment by fonsecas72** [comment]
@kevcodez It should work. Go and try the branch I've mentioned. You should then have checkboxes (similar to github) whose should allow you to build whatever you want.

**Comment by tknerr** [comment]
@stephenc any update / approximate ETA on your [earlier comment](https://github.com/jenkinsci/bitbucket-branch-source-plugin/pull/25#issuecomment-286839729) regarding the cleanup?

**Comment by nickbroon** [comment]
:+1:

**Comment by sasah** [comment]
+1!!!

**Comment by casz** [comment]
It is being worked on through this EPIC: https://issues.jenkins-ci.org/browse/JENKINS-43426

**Comment by stephenc** [comment]
Here is a preview of the cleanup in the GitHub Branch Source. I am currently finalizing that clean-up and then I will be refactoring to make this easy to share commonality between the two plugins:

jenkins-43507

There is progress and hopefully this will explain why I have been pushing back on the checkboxes

**Comment by nRon** [comment]
Nice preview, the option "Only branches that are not also filed as PRs" would work excellent for our use case! We are standing by to upgrade asap :)

**Comment by szhem** [comment]
Looking forward for merge for about a year :)

**Comment by tknerr** [comment]
@stephenc status ping, anything ready enough we could take for a test ride?

**Comment by stephenc** [comment]
@tknerr getting close... sometime next week I hope

**Comment by tknerr** [comment]
Awesome, thanks @stephenc! Looking forward to give it a spin.

Am 25.05.2017 05:15 schrieb "Stephen Connolly" <notifications@github.com>:

**Comment by palfrey** [comment]
I'm trying to use @fonsecas72's version of this (as commented further up) and I'm hitting some problems:
```
ERROR: [Fri Jun 02 11:39:45 BST 2017] Could not fetch branches from source 28422453-5718-445e-8fe2-2b89e20326b2
java.lang.NullPointerException
	at com.cloudbees.jenkins.plugins.bitbucket.PullRequestSCMRevision.equals(PullRequestSCMRevision.java:44)
	at jenkins.branch.MultiBranchProject$SCMHeadObserverImpl.observe(MultiBranchProject.java:1918)
	at com.cloudbees.jenkins.plugins.bitbucket.BitbucketSCMSource.observeFactory(BitbucketSCMSource.java:531)
	at com.cloudbees.jenkins.plugins.bitbucket.BitbucketSCMSource.observe(BitbucketSCMSource.java:511)
	at com.cloudbees.jenkins.plugins.bitbucket.BitbucketSCMSource.retrievePullRequests(BitbucketSCMSource.java:452)
	at com.cloudbees.jenkins.plugins.bitbucket.BitbucketSCMSource.retrieve(BitbucketSCMSource.java:396)
	at jenkins.scm.api.SCMSource._retrieve(SCMSource.java:300)
	at jenkins.scm.api.SCMSource.fetch(SCMSource.java:210)
	at jenkins.branch.MultiBranchProject.computeChildren(MultiBranchProject.java:634)
	at com.cloudbees.hudson.plugins.folder.computed.ComputedFolder.updateChildren(ComputedFolder.java:219)
	at com.cloudbees.hudson.plugins.folder.computed.FolderComputation.run(FolderComputation.java:141)
	at jenkins.branch.MultiBranchProject$BranchIndexing.run(MultiBranchProject.java:973)
	at hudson.model.ResourceController.execute(ResourceController.java:98)
	at hudson.model.Executor.run(Executor.java:410)
```

Also, we'd like to build only PRs *and* the master branch. Is there any way to do this? So far, I can manage to get it to build PRs *or* build master, but no option for "only PRs plus regex-matched branches"

**Comment by DoDoENT** [comment]
I have the same use-case as @palfrey. Currently I have that implemented as custom groovy script which simply does not run the build actions if not on master branch or on PR.

**Comment by batesyuk** [comment]
Same for me; `master` and `beta` should build and deploy on every change, other than that it should ignore all other branches and unit test PRs

**Comment by stephenc** [comment]
FYI my refactoring is nearly ready. Will add merge commit builds for PRs, ability to not build origin branches that are files as a PR, lots of control over git extensions, etc... may even change the baby's nappy.

I will be looking for beta testers next week some time

**Comment by palfrey** [comment]
I started doing some work on this and if this is of any use, https://github.com/lshift/bitbucket-branch-source-plugin/tree/jenkins-36283 has 
* Fixes to the exception I saw earlier
* Support for PRs + regex-matched branch builds
* Sets a dummy username/email for Jenkins in Git before trying to locally merge branches so that type of build works again
* Master merged in

**Comment by stephenc** [comment]
Traits makes most of that irrelevant. The git behaviours can be attached directly. More advanced "custom" behaviours can be encapsulated in their own plugins so that only those that need the functionality clutter their UI... and only when using it

**Comment by stephenc** [comment]
superseded by #53

**Pull request closed without merging**